lib/IDE/CompletionLookup.cpp

```diff
@@ -2017,9 +2021,10 @@ void CompletionLookup::foundDecl(ValueDecl *D, DeclVisibilityKind Reason,
 
     if (auto *GP = dyn_cast<GenericTypeParamDecl>(D)) {
       addGenericTypeParamRef(GP, Reason, dynamicLookupInfo);
-      for (auto *protocol : GP->getConformingProtocols())
-        addConstructorCallsForType(protocol->getDeclaredInterfaceType(),
-                                   GP->getName(), Reason, dynamicLookupInfo);
+      auto type = GP->getDeclContext()->mapTypeIntoContext(
+          GP->getDeclaredInterfaceType());
+      addConstructorCallsForType(type, GP->getName(), Reason,
+                                 dynamicLookupInfo);
```

Comment:
@slavapestov Previously, I was doing
```cpp
auto type = CurrDeclContext->mapTypeIntoContext(GP->getDeclaredInterfaceType())
```
But it didn't work (crashed) for cases like:
```swift
protocol P {
  struct S {
    func foo() { <HERE> }
  }
}
```
It was crashing when `GP` is `Self` for `P`. In this case, I believe `CurrDeclContext` (`S.foo()`) doesn't have a valid generic environment because it's invalid anyway. So I updated it to:
```cpp
auto type = GP->getDeclContext()->mapTypeIntoContext(GP->getDeclaredInterfaceType())
```
Do you think this is safe? Inside `addConstructorCallsForType()`, We just `swift::lookupSemanticMember(DC, type, DeclBaseName::createConstructor())`.

Reply:
The generic parameter `Self` of `protocol P` is not visible from inside `struct S`. Why was it showing up here?

Reply:
Hmm, maybe `lookupVisibleDecls()` is not working properly.
Do you mean `Self` of `protocol P` should be shadowed by `Self` of `struct S`? Or it should not be visible anyways?

Reply:
It's not visible. Generic parameters are only visible from nested lexical scopes and never via protocol or superclass inheritance. It is also true that `Self` inside of a concrete type means that concrete type, but the previous fact is true of all generic parameters not just `Self`.

Reply:
Now if you have a member of the protocol that is visible via the concrete type, eg

```swift
protocol P {
  func foo() -> Self
}

struct S : P {
  func bar() {
    ... completion here ...
  }
}
```

Then the interface type of `P.foo` is `(Self) -> () -> Self`. To get the type of the reference from inside `S.bar()`, you must apply a substitution:

```c++
auto fooType = fooDecl->getInterfaceType();
auto barSelfType = barDecl->getDeclContext()->getSelfInterfaceType();
auto substFooType = fooType->subst(barSelfType->getContextSubstitutionMap(fooDecl->getDeclContext()));
```

Now `substFooType` is going to be `(S) -> () -> S`.

Reply:
OK, thank you for explanation. I will look into `lookupVisibleDecls()`.
If it doesn't emit generic parameter of outer types (including `Self`), should we use `CurDeclContext` for `mapTypeIntoContext`?

Reply:
It will report generic parameters of outer types, eg

```swift
struct Outer<T> {
  struct Inner<U> {
    // both T and U visible here
  }
}
```

In this case, `innerDC->mapTypeIntoContext(tDecl->getDeclaredInterfaceType())` will work.

What it shouldn't find is generic parameters of superclass types and protocols that each type conforms to.

Reply:
Ah, I this is not a superclass or conforming protocols issue.
```
protocol P {
  struct S {
    func foo() { <HERE> }
  }
}
```
`struct S` is just erroneously nested in a `protocol P`. It's invalid anyway, but I just want to avoid the crash.

> ```struct Outer<T> {
>  struct Inner<U> {
>    // both T and U visible here
>  }
> }
> ```
:+1:

Reply:
Ah, this is why DeclContext::getParentForLookup() exists. When starting from inside S, lookupVisibleDecls() should be using getParentForLookup() instead of getParent(). This way it will never visit P and find the `Self` parameter erroneously.

Reply:
`getParentForLookup()` in lookupVisibleDecls() seems to work. Thank you!